modules/light-clients/06-solomachine/update.go

```diff
@@ -0,0 +1,141 @@
+	// assert update sequence is current sequence
+	if header.Sequence != cs.Sequence {
+		return sdkerrors.Wrapf(
+			clienttypes.ErrInvalidHeader,
+			"header sequence does not match the client state sequence (%d != %d)", header.Sequence, cs.Sequence,
+		)
+	}
+
+	// assert update timestamp is not less than current consensus state timestamp
+	if header.Timestamp < cs.ConsensusState.Timestamp {
+		return sdkerrors.Wrapf(
+			clienttypes.ErrInvalidHeader,
+			"header timestamp is less than to the consensus state timestamp (%d < %d)", header.Timestamp, cs.ConsensusState.Timestamp,
+		)
+	}
+
+	// assert currently registered public key signed over the new public key with correct sequence
+	headerData := &HeaderData{
+		NewPubKey:      header.NewPublicKey,
+		NewDiversifier: header.NewDiversifier,
+	}
+
+	dataBz, err := cdc.Marshal(headerData)
+	if err != nil {
+		return err
+	}
+
+	signBytes := &SignBytes{
+		Sequence:    header.Sequence,
+		Timestamp:   header.Timestamp,
+		Diversifier: cs.ConsensusState.Diversifier,
+		Path:        []byte{},
+		Data:        dataBz,
+	}
+
+	data, err := cdc.Marshal(signBytes)
+	if err != nil {
+		return err
+	}
+
+	sigData, err := UnmarshalSignatureData(cdc, header.Signature)
+	if err != nil {
+		return err
+	}
+
+	publicKey, err := cs.ConsensusState.GetPubKey()
+	if err != nil {
+		return err
+	}
+
+	if err := VerifySignature(publicKey, data, sigData); err != nil {
+		return sdkerrors.Wrap(ErrInvalidHeader, err.Error())
+	}
+
+	return nil
+}
+
+func (cs ClientState) verifyMisbehaviour(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, misbehaviour *Misbehaviour) error {
+	// NOTE: a check that the misbehaviour message data are not equal is done by
+	// misbehaviour.ValidateBasic which is called by the 02-client keeper.
+	// verify first signature
+	if err := cs.verifySignatureAndData(cdc, misbehaviour, misbehaviour.SignatureOne); err != nil {
+		return sdkerrors.Wrap(err, "failed to verify signature one")
+	}
+
+	// verify second signature
+	if err := cs.verifySignatureAndData(cdc, misbehaviour, misbehaviour.SignatureTwo); err != nil {
+		return sdkerrors.Wrap(err, "failed to verify signature two")
+	}
+
+	return nil
+}
+
+// UpdateState updates the consensus state to the new public key and an incremented sequence.
+// A list containing the updated consensus height is returned.
+func (cs ClientState) UpdateState(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, clientMsg exported.ClientMessage) []exported.Height {
+	smHeader, ok := clientMsg.(*Header)
+	if !ok {
+		panic(fmt.Errorf("unsupported ClientMessage: %T", clientMsg))
+	}
+
+	// create new solomachine ConsensusState
+	consensusState := &ConsensusState{
+		PublicKey:   smHeader.NewPublicKey,
+		Diversifier: smHeader.NewDiversifier,
+		Timestamp:   smHeader.Timestamp,
+	}
+
+	cs.Sequence++
+	cs.ConsensusState = consensusState
+
+	clientStore.Set(host.ClientStateKey(), clienttypes.MustMarshalClientState(cdc, &cs))
```

Comment:
we should have a `setClientState` function

Reply:
#1887

Reply:
We have one in [`client_state.go`](https://github.com/cosmos/ibc-go/blob/updated-02-client-refactor/modules/light-clients/06-solomachine/client_state.go#L261). 

My bad on forgetting to use it here!

Reply:
https://github.com/cosmos/ibc-go/pull/1891